### PR TITLE
[WIP] Update foreman inventory to use foreman's inventory report

### PR DESCRIPTION
**Pull request by sjha4**
##### SUMMARY
Currently the foreman inventory script relies on API calls to foreman to fetch hosts, details and facts. This is achieved my making one call to get the list of hosts, followed by 2 calls per host to fetch details and facts. Unsurprisingly, this is slow and sometimes take hours to complete.

Proposed change is to implement a report which provides all the required data from foreman (WIP here: https://github.com/theforeman/foreman_ansible/pull/299) Following this, the inventory scripts needs to API calls. First to generate the report and the second to fetch the generated data.

##### ISSUE TYPE
 - Bugfix Pull Request

##### AWX VERSION
```
awx: 7.0.0
```

### CODE REVIEW

**Comment by wenottingham** [comment]
This would need to go to https://github.com/ansible/ansible/ first.

**Comment by sjha4** [comment]
@wenottingham : Thanks for the pointer :+1: I have opened a request in ansible here: https://github.com/ansible/ansible/pull/62438

File: awx/plugins/inventory/foreman.py [hunk 1]
```diff
@@ -203,6 +203,19 @@ def _get_json(self, url, ignore_errors=None, params=None):
...
+        url = "%s/ansible/api/v2/ansible_inventories/schedule_report" % self.foreman_url
+        session = self._get_session()
+        params = {'input_values': {'required':['all']}}
+        ret = session.post(url, json=params)
+        url = "{0}/api/v2/report_templates/153-Ansible%20Inventory/report_data/{1}".format(self.foreman_url, ret.json().get('job_id'))
+        while True:
+            ret = session.get(url)
+            if ret.status_code != 204:
+                break
+            sleep(5)
```
Comment (AlanCoding) [inline awx/plugins/inventory/foreman.py]: I'm going to have question the reason for this sleep

**Comment by AlanCoding** [comment]
at the linked Ansible PR we have the statement

> We are no longer accepting changes to the contrib inventory scripts.

Maybe this becomes the source of truth for the script, but I'm not totally sure. Since there is an inventory plugin for this, the actual logic can still be reviewed there. That's about the best proposal I have right now.

**Comment by ws4e** [comment]
> This would need to go to https://github.com/ansible/ansible/ first.

The problem is that that one is quite a bit older and less maintained than the foreman inventory plugin currently in the AWX project.  If the foreman inventory plugin in the ansible CLI project is going to become the main source, then it needs to first come up to speed with the AWX plugin version, which is newer, has extra features like the ansible_ssh_host thing I use, that is not present in the ansbile CLI plugin version.   

Just telling people to first go to the CLI version is probably not the right thing to do until that version is at least as new as the one currently in the AWX inventory plugins.

File: awx/plugins/inventory/foreman.py (hunk 1 above)
Comment (sjha4) [reply awx/plugins/inventory/foreman.py]: The change we are making with this PR is that we are introducing a choice for tower to fetch inventory by generating an inventory report in foreman. The flow is:
1. Call API to start generating report in foreman.
2. Step 1 returns the job id which we poll at fixed intervals to check when it's complete.
While the job is running in foreman, the API returns status 204. 
3. Once the job completes in foreman, we get a response in our next call to the API and we move ahead to process it.

**Comment by ryanpetrello** [comment]
@sjha4 can you rebase this PR against `devel` and re-open it?  It's very far behind (it currently has no chance of passing CI in this state, and it's using up our public CI resources).

**Comment by sjha4** [comment]
@ryanpetrello : I don't seem to have access to re-open this. I have opened this https://github.com/ansible/awx/pull/4988 to run CI but I can rebase here if needed.

**Pull request closed without merging**